constraints/sidecar-injection-annotation.yaml

```diff
@@ -1,14 +1,14 @@
 apiVersion: constraints.gatekeeper.sh/v1beta1
-kind: PodSidecarInjection
+kind: SidecarInjectionAnnotation
 metadata:
   name: pod-sidecar-injection
 spec:
   enforcementAction: deny
   match:
     kinds:
     - apiGroups:
-      - ""
+      - apps
       kinds:
-      - Pod
+      - Deployment
```

Comment:
The best practice here is actually matching `Pod` only. Matching `Pod` and `Deployment` in a single ConstraintTemplate can cause some problems. The reason why I switched to `Deployment` is so that we can detect non-compliant `Deployment` during CI.

Instead of using `Deployment`, what do you think of switching it to `Pod` for demo purposes? I know that the best practice here is to use a `Deployment` to generate `Pod`  but with Pod-based constraint, we can only catch non-compliant Pod in an actual cluster, not in a CI environment.

Have you tried running `kpt eval fn` when you add `sidecar.istio.io/inject: "false"` to the pod template of the deployment? Was it able to detect the violation?

Reply:
Yep, I tried, but it won't work because it's a `Deployment` resource.

In an other project, I was able to make this working like this for the template:
```
violation[{"msg": msg}] {
    input.review.kind.kind == "Pod"
    pod := input.review.object
    pod.metadata.annotations[forbidden_injection_annotation["key"]] == forbidden_injection_annotation["value"]
    msg := sprintf("The annotation %v: %v should not be applied on workload pods", [forbidden_injection_annotation["key"], forbidden_injection_annotation["value"]])
}
violation[{"msg": msg}] {
    input.review.kind.kind == "Deployment"
    pod := input.review.object.spec.template
    pod.metadata.annotations[forbidden_injection_annotation["key"]] == forbidden_injection_annotation["value"]
    msg := sprintf("The annotation %v: %v should not be applied on workload pods", [forbidden_injection_annotation["key"], forbidden_injection_annotation["value"]])
}
```
Thoughts?

Reply:
We are actually working on addressing this exact issue upstream on handling pod violations within the spec of pod-generating resources like Deployment without creating duplicated rego. Since there is no good way to do so at this moment, I am okay with using the rego above.